### PR TITLE
Http request predicate option

### PR DESCRIPTION
**Pull request by worldspawn**
This adds the functionality discussed here - https://github.com/serilog/serilog/issues/554

### CODE REVIEW

**Comment by merbla** [comment]
@worldspawn just looking at #21, seems like a bit of an overlap here.  Would some consolidation of the two cover this?

cc:// @enkafan @nblumhardt

**Comment by worldspawn** [comment]
Our two respective pull requests cover different things. Mine is about whether any logging should happen at all and #21 is about adding request body data to that log. Similar but not the same.

**Comment by merbla** [comment]
No problem, was thinking an option of predicates at the higher object level  e.g. `HttpContextCurrent`

**Comment by worldspawn** [comment]
Oh yeah I guess that has more utility. I was very focused on the _request_ when i wrote this. I'll make that change sometime this weekend. Both PRs have there seperate uses though.

A service resolver could clean up a lot of this stuff, the application could ask the container for all ILogRequestPredicates or all ILogFormDataPredicates and that would remove the need to add all these predicate properties to the class. Anything registered in the resolver container would be what is used.

The main serilog package also could benefit from such an approach. Currently there are things that cannot be extend such as the scalar value destructures because theres no extensibility point.

**Comment by merbla** [comment]
@worldspawn be keen to see how this one falls out

**Comment by worldspawn** [comment]
I've retargeted the predicates to the HttpContext.

File: src/SerilogWeb.Classic/Classic/ApplicationLifecycleModule.cs [hunk 1]
```diff
@@ -36,6 +37,7 @@ public class ApplicationLifecycleModule : IHttpModule
         static volatile IEnumerable<string> _filteredKeywords = new[] { "password" };
         static volatile LogEventLevel _requestLoggingLevel = LogEventLevel.Information;
         static volatile LogEventLevel _formDataLoggingLevel = LogEventLevel.Debug;
+        static volatile IList<Func<HttpContext, bool>> _requestPredicates = new List<Func<HttpContext, bool>>();
```
Comment (nblumhardt) [inline src/SerilogWeb.Classic/Classic/ApplicationLifecycleModule.cs:40]: If we're going to allow on-the-fly adding of predicates, needs to be something like a `ConcurrentBag<>`.

File: src/SerilogWeb.Classic/Classic/ApplicationLifecycleModule.cs [hunk 2]
```diff
@@ -36,6 +37,7 @@ public class ApplicationLifecycleModule : IHttpModule
         static volatile IEnumerable<string> _filteredKeywords = new[] { "password" };
         static volatile LogEventLevel _requestLoggingLevel = LogEventLevel.Information;
         static volatile LogEventLevel _formDataLoggingLevel = LogEventLevel.Debug;
```
Comment (nblumhardt) [inline src/SerilogWeb.Classic/Classic/ApplicationLifecycleModule.cs:39]: I think the number of memory barriers these are generating is getting a bit out of hand, perhaps we drop all the `volatile`s and just accept a very rare chance of "delayed action" when configuring these (almost impossible in ASP.NET apps).

File: src/SerilogWeb.Classic/Classic/ApplicationLifecycleModule.cs [hunk 3]
```diff
@@ -69,6 +71,16 @@ public static void Register()
         }
 
         /// <summary>
+        /// You can add predicates to this list and they will be evaluated before
+        /// logging the request. If *any* fail the request will not be logged.
+        /// </summary>
+        public static IList<Func<HttpContext, bool>> RequestPredicates
+        {
+            get { return _requestPredicates; }
+            set { _requestPredicates = value; }
```
Comment (nblumhardt) [inline src/SerilogWeb.Classic/Classic/ApplicationLifecycleModule.cs:80]: Probably no need for `set` here.

File: src/SerilogWeb.Classic/Classic/ApplicationLifecycleModule.cs [hunk 4]
```diff
@@ -155,7 +167,7 @@ public void Init(HttpApplication application)
                     stopwatch.Stop();
 
                     var request = HttpContextCurrent.Request;
-                    if (request == null)
+                    if (request == null || !(_requestPredicates ?? Enumerable.Empty<Func<HttpContext, bool>>()).All(p => p(application.Context)))
```
Comment (nblumhardt) [inline src/SerilogWeb.Classic/Classic/ApplicationLifecycleModule.cs:170]: Could we try:

``` csharp
if (request == null || _requestPredicates.Count != 0 && _requestPredicates.Any(p => !p(application.Context)))
```

Saves a closure allocation/iteration if no predicates are configured.

**Comment by worldspawn** [comment]
hi, i'll try and address the feedback on the weekend. thx

**Comment by phil-scott-78** [comment]
Ran across the need for this today - a new signalr implementation is blowing up the logs. @worldspawn I'm happy to take it to completion if you haven't had the time to loop back around to this.

**Pull request merged**